### ISSUE TITLE
Ctrl-C terminates FAR in various Terminal emulators

### ISSUE THREAD

**Issue by jhhcs** [FarGroup/FarManager#442]
I have a problem with FAR that I first noticed when using Windows Terminal: https://github.com/microsoft/terminal/issues/5944

Today, I attempted to use FAR with [ConsoleZ](https://github.com/cbucher/console) and the same problem surfaced. On my machine, the problem is reproduced as follows:

- Start FAR inside ConsoleZ
- run `ping -t 1.1.1.1` (or something else that runs for a bit)
- Press <kbd>Ctrl</kbd>+<kbd>C</kbd>

The result for me is that the entire FAR process terminates, while I expected to only terminate the running command. The guys over at Microsoft were not able to reproduce my problem, but another FAR user I know can reproduce it on their machine as well and I can also reproduce it in a Windows 10 Virtual Machine. I am using Windows 10 19042.1165.

I was hoping that someone here might have an idea of what is going on, and maybe something can be adjusted in FAR to prevent this problem.

**Comment by alabuzhev** [comment]
@jhhcs I'm relatively sure that Far handles Ctrl+C correctly.
However, there are plugins and they can do anything. Including breaking that handling.
Please try without any plugins (far /p). If that helps, try removing them one by one to find the culprit.

**Comment by huettenhain** [comment]
Hey! I wanted to make sure and followed your advice, but it still happens. The following is from a fresh Windows Sandbox session, the only things installed are ConsoleZ and FAR, and I have removed the Plugins folder (as you can see):

![recording2](https://user-images.githubusercontent.com/11869704/132073690-0f0199a9-c18a-449c-9066-cc2af46f05a6.gif)

**Comment by huettenhain** [comment]
PS: running `far /p` does not change anything either, but I suppose that is the same as removing the plugins folder anyway.

**Comment by yegor-mialyk** [comment]
I'm using Far with Windows Terminal (latest version) from time to time and I cannot reproduce this.

Far 3.0.5883.0 / Windows 10.0.19043.1165 (10) and Windows 10.0.22000.176 (11)

**Comment by jhhcs** [comment]
By the way, I am not saying that FAR handles anything incorrectly. I have absolutely no idea where the bug is coming from. I am only hoping someone here can reproduce it and help me figure out the root cause.

**Comment by alabuzhev** [comment]
[Here](https://github.com/microsoft/terminal/issues/5944#issuecomment-885597703) you were saying that you can't reproduce it in a virtual environment, but looks like you can now. Has anything changed recently?

Also, you mentioned some "QObject" messages there. QObject is a part of Qt framework, which is definitely not used in Far, and neither in Windows (as far as I'm aware), so it must be a 3rd party dll injected into the process.
Please try to capture the list of loaded modules using `tasklist /fi "ImageName eq far.exe" /m` or process explorer (Ctrl+L, Ctrl+D).
Alternatively, try looking for the `QObject` string in dlls on your C drive.

**Comment by alabuzhev** [comment]
Also, in the video you're starting far with just `far` command from an unrelated directory in cmd. That shouldn't work out of the box. Is it your own script / something else?

**Comment by alabuzhev** [comment]
If you can reproduce it in an environment with a debugger installed, the easiest way to understand what is going on could be attaching it to the process, pressing ctrl+c and skipping the initial DBG_CONTROL_C exception.

**Comment by jhhcs** [comment]
What has changed is only that I use ConsoleZ rather than Windows Terminal. The reason I can start far from the shell is simply because `far.exe` is in my `PATH`. And let me emphasize again, the above demonstration is from a fresh sandbox instance with _only_ FAR and ConsoleZ installed, I don't really believe that anything is messing with FAR or injecting stuff into it. I don't think the `QObject` related error messages have anything to do with this either, they only occurred once, and that was not in a clean sandbox environment. Nevertheless, I will try your suggestions and report back.

**Comment by huettenhain** [comment]
So FAR does receive the <kbd>Ctrl</kbd>+<kbd>C</kbd> after the ping has already terminated:

![image](https://user-images.githubusercontent.com/11869704/132093317-b3b73559-06e2-49c4-909e-d59d3ac101d5.png)

I do not know if FAR is intended to receive this signal or not. No specifically weird modules are loaded at the time:

![image](https://user-images.githubusercontent.com/11869704/132093331-22ad805b-0fbe-43d4-b0d0-546ab9c7b190.png)

And then FAR just dies from the <kbd>Ctrl</kbd>+<kbd>C</kbd>, i.e. taking another step in the debugger just flatlines the process.

**Comment by alabuzhev** [comment]
Yes, it is intended to receive the signal and ignore it:
https://github.com/FarGroup/FarManager/blob/f97c66bfc61abfbeffbaafdfebd279fca0dcd758/far/interf.cpp#L174-L180

![image](https://user-images.githubusercontent.com/11453922/132093814-8a195415-41a4-4c48-a472-0958d5cb425a.png)

It is a little strange that .Net stuff is loaded into the process.

**Comment by alabuzhev** [comment]
> taking another step in the debugger just flatlines the process

If you see no other exceptions, something is probably flatlines it manually.
Try putting breakpoints on ExitProcess and TerminateProcess functions.

**Comment by yegor-mialyk** [comment]
I would suggest performing the following experiment:

1. Open your ConsoleZ (it opens "cmd" by default)
2. Run “powershell”
3. Run “ping -t”
4. Press Ctrl+C

In other words, ConsoleZ -> PowerShell -> Ping -> Ctrl+C

Did it return you back to PowerShell or cmd?

**Comment by huettenhain** [comment]
@yegor-mialyk I did try that, and it returns to PowerShell as expected, PowerShell does not terminate as FAR does.

**Comment by huettenhain** [comment]
Oh for crying out loud. I am so sorry guys, I figured it out, and of course it's _not_ a problem with FAR. I had installed FAR and ConsoleZ using chocolatey, and I was **wrong**: It does not put `far.exe` on the `PATH`, it creates a shim executable which it then puts on the `PATH`. The shim executable seems to be the one that incorrectly handles the keyboard interrupt: Running `far.exe` directly from the installation directory does not have any of the issues. I was also wrong about ConsoleZ being the only change I made for the test: This was also the only time I used chocolatey for the test, but I did not think much of it. 

I am so sorry, I should have double-checked my assumptions earlier.
